### PR TITLE
Update controller packages to match core v42

### PR DESCRIPTION
**Pull request by Gudahtt**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

Most controller packages have been updated to the versions included in the core monorepo v42 release. The keyring controller was the only package held back, due to the BigInt incompatibility.

The breaking change for most of these updates was related to the removal of the `isomorphic-fetch` package. That package would polyfill the `fetch` API if it wasn't present. It's not a breaking change for mobile because mobile already includes its own polyfill for the `fetch` API.

The only other breaking change is to the gas fee controller, which has made a constructor parameter required rather than optional. This does not affect mobile because mobile already sets that parameter..

See here for the release notes for each updated controller: https://github.com/MetaMask/core/releases/tag/v42.0.0

A number of packages had their ranges reduced from `^` to `~` or were pinned so that the scope of this PR could be kept reasonably small. Two required resolutions to avoid bringing in unwanted updates. Here are the details:

* `@metamask/controller-utils` range set to `~3.0.0` with a resolution
  Later versions of this package use the @metamask/utils package, which has introduced BigInt-related problems for mobile recently. We can remove this resolution in a future PR after carefully considering the impact, or after the React Native upgrade.
* `@metamask/approval-controller` range set to `~2.0.0` with a resolution
  This was necessary to avoid dependencies upon later `@metamask/controller-utils` versions
* `"@metamask/assets-controllers` pinned to `5.0.0`  
  Avoids dependency upon `@metamask/network-controller@6.0.0`, which is required by `@metamask/assets-controllers@5.0.1`
* `@metamask/gas-fee-controller` pinned to `4.0.0`
  Avoids dependency upon `@metamask/network-controller@6.0.0`, which is required by `@metamask/gas-fee-controller@4.0.1`
* `@metamask/permission-controller` range reduced to `~3.0.0`
  Avoids dependency upon later versions of `@metamask/approval-controller`
* `@metamask/transaction-controller` pinned to `4.0.0`
  Avoids dependency upon `@metamask/network-controller@6.0.0`, which is required by `@metamask//transaction-controller@4.0.1`

**Issue**

Progresses https://github.com/MetaMask/mobile-planning/issues/798

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

### CODE REVIEW

**Comment by Gudahtt** [comment]
This depends upon https://github.com/MetaMask/metamask-mobile/pull/6124

File: package.json [hunk 1]
```diff
@@ -142,26 +142,26 @@
...
+    "@metamask/base-controller": "^2.0.0",
+    "@metamask/composable-controller": "^2.0.0",
     "@metamask/contract-metadata": "^2.1.0",
-    "@metamask/controller-utils": "^2.0.0",
+    "@metamask/controller-utils": "^3.0.0",
     "@metamask/design-tokens": "^1.11.1",
     "@metamask/eth-sig-util": "^4.0.1",
     "@metamask/etherscan-link": "^2.0.0",
-    "@metamask/gas-fee-controller": "^3.0.0",
+    "@metamask/gas-fee-controller": "4.0.0",
```
Comment (Gudahtt) [inline package.json:155]: Temporarily pinned to reduce the scope of this PR

File: package.json [hunk 2]
```diff
@@ -142,26 +142,26 @@
...
-    "@metamask/phishing-controller": "^2.0.0",
+    "@metamask/message-manager": "^2.0.0",
+    "@metamask/network-controller": "^5.0.0",
+    "@metamask/permission-controller": "^3.0.0",
+    "@metamask/phishing-controller": "^3.0.0",
     "@metamask/preferences-controller": "^2.1.0",
     "@metamask/sdk-communication-layer": "0.2.2",
     "@metamask/swaps-controller": "^6.8.0",
-    "@metamask/transaction-controller": "^3.0.0",
+    "@metamask/transaction-controller": "4.0.0",
```
Comment (Gudahtt) [inline package.json:164]: Temporarily pinned to reduce the scope of this PR

File: package.json [hunk 3]
```diff
@@ -113,6 +113,8 @@
     "react-native-level-fs/**/bl": "^1.2.3",
     "react-native-level-fs/**/semver": "^4.3.2",
     "@metamask/contract-metadata": "^2.1.0",
+    "@metamask/controller-utils": "~3.0.0",
+    "@metamask/approval-controller": "~2.0.0",
```
Comment (Gudahtt) [inline package.json:117]: This has been kept at ~3 to avoid dependencies upon later versions of the controller-utils package.

File: package.json [hunk 4]
```diff
@@ -142,26 +144,26 @@
...
-    "@metamask/gas-fee-controller": "^3.0.0",
+    "@metamask/gas-fee-controller": "4.0.0",
     "@metamask/keyring-controller": "^1.0.1",
-    "@metamask/message-manager": "^1.0.2",
-    "@metamask/network-controller": "^4.0.0",
-    "@metamask/permission-controller": "^2.0.0",
-    "@metamask/phishing-controller": "^2.0.0",
+    "@metamask/message-manager": "^2.0.0",
+    "@metamask/network-controller": "^5.0.0",
+    "@metamask/permission-controller": "~3.0.0",
```
Comment (Gudahtt) [inline package.json:161]: This has been kept at ~3 to avoid dependencies upon later versions of the approval controller

**Comment by Gudahtt** [comment]
The Bitrise iOS e2e tests succeeded: https://app.bitrise.io/build/3abfec4a-1077-45c6-89c1-0a3872a319e9
Android failed; still looking into that

File: package.json [hunk 5]
```diff
@@ -113,6 +113,8 @@
     "react-native-level-fs/**/bl": "^1.2.3",
     "react-native-level-fs/**/semver": "^4.3.2",
     "@metamask/contract-metadata": "^2.1.0",
+    "@metamask/controller-utils": "~3.0.0",
```
[the remark this answers is not in the document]
Comment (sethkfman) [reply package.json:116]: Perfect! Would you mind adding a comment on the same line so we can make sure to update it once RN upgrade goes in?

File: package.json (hunk 3 above)
Comment (sethkfman) [reply package.json:117]: Same comment as above. Thx

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply package.json:116]: `package.json` does not allow comments unfortunately

**Comment by sethkfman** [comment]
@Gudahtt I sent over a few comments. Once you tracked down the results on the Android side let me know and I can review again.

**Comment by Gudahtt** [comment]
Got a successful Android e2e test run here (passing all tests except 2 that are failing on `main`): https://app.bitrise.io/build/e87c3485-4efe-4596-8f27-83851669eae7

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply package.json:116]: I have added all of these details to the PR description so that they're easier to find for anyone reading this in the future

**Review by sethkfman — APPROVED** [review]
LGTM

**Pull request merged**